src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -516,9 +644,9 @@ private bool TryCreateEntityReferenceSegment(string text)
             this.parsedSegments.Add(navPropLinkSegment);
 
             // Add key segments back to the parsed segments
-            while (keySegmentsForPreviousPathSegment.Count > 0)
+            while (keyAndFilterSegmentsForPreviousPathSegment.Count > 0)
             {
-                this.parsedSegments.Add(keySegmentsForPreviousPathSegment.Pop());
+                this.parsedSegments.Add(keyAndFilterSegmentsForPreviousPathSegment.Pop());

```

Comment:
Some of the assumptions above are not correct. To preserve backwards compatibility, revert the changes for the $ref and add a new ReferenceSegment to add corrected behavior.